From 1bf64ee6e25ff8256337bfe9c13ceaf27c06c2aa Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 1 Nov 2023 22:36:40 +0530
Subject: [PATCH 3/4] hostapd: fix simultaneous scanning failure in case of MLO

Currently only 1 scan can be performed on per phy level at a time
in mac80211. Due to this, if via other underlying radio, a scan
request is sent, kernel returns -EBUSY. And then hostapd would
try to set the interface in station mode if it was originally
in AP mode and retry sending a scan request. However, this
behaviour is expected in case of multi link operation and hence
there is no need to switch the mode as such.

Hence, add logic to not change the nl mode if kernel returns
-EBUSY during multi link AP operation. The caller can accordingly
decide and if needed, it can re-schedule a scan request after
some time.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 src/drivers/driver_nl80211_scan.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

--- a/src/drivers/driver_nl80211_scan.c
+++ b/src/drivers/driver_nl80211_scan.c
@@ -413,6 +413,40 @@ int wpa_driver_nl80211_scan(struct i802_
 		if (drv->hostapd && is_ap_interface(drv->nlmode)) {
 			enum nl80211_iftype old_mode = drv->nlmode;
 
+#ifdef CONFIG_IEEE80211BE
+			/* For Multi Link BSS, retry scan if any other links
+			 * are busy scanning
+			 */
+			if (ret == -EBUSY &&
+			    nl80211_link_valid(bss->valid_links, params->mlo_link_id)) {
+				struct i802_bss *link_bss;
+				u8 link_id;
+
+				wpa_printf(MSG_DEBUG,
+					   "nl80211: Scan trigger on Multi Link BSS failed (requested link=%d on interface %s)",
+					   params->mlo_link_id, bss->ifname);
+
+				for (link_bss = drv->first_bss; link_bss;
+				     link_bss = link_bss->next)
+					if (link_bss->scan_link)
+						break;
+
+				if (!link_bss) {
+					wpa_printf(MSG_DEBUG,
+						   "nl80211: Interface information already running scan not available");
+					goto fail;
+				}
+
+				link_id =
+				       nl80211_get_link_id_from_link(link_bss,
+								     link_bss->scan_link);
+				wpa_printf(MSG_DEBUG,
+					   "nl80211: Scan already running on interface %s link %d",
+					   link_bss->ifname, link_id);
+				goto fail;
+			}
+#endif /* CONFIG_IEEE80211BE */
+
 			/*
 			 * mac80211 does not allow scan requests in AP mode, so
 			 * try to do this in station mode.
